File: Leetcode First 200 JavaScript solution/121. Best time to buy and sell stock.js
var maxProfit = function(prices) {

    if (prices.length === 0) return 0;
    let profit = 0, buy = prices[0];

    for (let price of prices) {
        const diff = price - buy;
        profit = Math.max(diff, profit);
        buy = Math.min(buy, price);
    }
    return profit;
}

// o(n) solution here!

File: Leetcode First 200 C++ solution/121. Best Time to Buy and Sell Stock.cpp
#include <vector>
using namespace std;

class Solution {
    public:
        int maxProfit(vector<int>& prices) {
            if (prices.size() == 0) return 0;
            int profit = 0, buy = prices[0];
            for (auto price : prices) {
                int diff = price - buy;
                profit = max(profit, diff);
                buy = min(buy, price);
            }
            return profit;
        }
};

// use the greedy solution to solve this question
// get the maximum profit each time
// need to memorize the smallest buy in price
// o(n)
